### PR TITLE
Port pagination to replychat

### PR DESCRIPTION
**Pull request by greatbacon**
This PR utilizes the paginated query action inside the ReplyChat view.

It will expand the paginated query action to also include ties related to the page items returned by the query

### CODE REVIEW

File: src/lib/infra/deploy.task.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ export const task: Task = {
 		log.debug(`Working with artifact: ${artifactName}`);
 		unwrap(
 			await spawn('tar', [
-				'-cvf',
+				'-cf',
```
Comment (ryanatkn) [inline src/lib/infra/deploy.task.ts:27]: # 🥂

File: src/lib/plugins/greatbacon/reply-chat/ReplyChat.svelte [hunk 2]
```diff
@@ -1,40 +1,41 @@
...
 	import {getViewContext} from '$lib/vocab/view/view';
 	import TextInput from '$lib/ui/TextInput.svelte';
 	import Mention from '$lib/plugins/feltcoop/mention/Mention.svelte';
-	import {sortEntitiesByCreated} from '$lib/vocab/entity/entityHelpers';
 	import type {Entity} from '$lib/vocab/entity/entity';
 	import {slide} from 'svelte/transition';
+
 	import {lookupPersona} from '$lib/vocab/actor/actorHelpers';
+	import {createPaginatedQuery} from '$lib/util/query';
+	import PendingButton from '@feltjs/felt-ui/PendingButton.svelte';
```
Comment (ryanatkn) [inline src/lib/plugins/greatbacon/reply-chat/ReplyChat.svelte:16]: some cleanup here, moving `import {slide} from 'svelte/transition';` and `import PendingButton from '@feltjs/felt-ui/PendingButton.svelte';` up to the top section and deleting the space that creates a 3rd section

File: src/lib/plugins/feltcoop/chat/Chat.svelte [hunk 3]
```diff
@@ -17,12 +17,15 @@
 	let text = '';
 
 	$: shouldLoadEntities = browser && $socket.open;
+
 	$: query = shouldLoadEntities
 		? createPaginatedQuery(ui, actions, {
 				actor: $persona.persona_id,
 				source_id: $space.directory_id,
+				sibling: 'none',
```
Comment (ryanatkn) [inline src/lib/plugins/feltcoop/chat/Chat.svelte:25]: what do you think about making this the default and making `sibling` optional? would match how we do `pageSize` and `pageKey`

File: src/lib/plugins/feltcoop/entity-explorer/EntityExplorer.svelte [hunk 4]
```diff
@@ -30,6 +30,7 @@
 		const result = await actions.ReadEntitiesPaginated({
 			actor: $persona.persona_id,
 			source_id: $space.directory_id,
+			sibling: 'none',
```
Comment (ryanatkn) [inline src/lib/plugins/feltcoop/entity-explorer/EntityExplorer.svelte:33]: would be able to delete this too if making optional

File: src/lib/ui/EntityTree.svelte [hunk 5]
```diff
@@ -33,6 +33,7 @@
 		const result = await actions.ReadEntitiesPaginated({
 			actor: $persona.persona_id,
 			source_id: $entity.entity_id,
+			sibling: 'none',
```
Comment (ryanatkn) [inline src/lib/ui/EntityTree.svelte:36]: also could delete here

File: src/lib/app/actionTypes.ts [hunk 6]
```diff
@@ -549,6 +553,7 @@ export interface ReadEntitiesPaginatedParams {
 	source_id: number;
 	pageSize?: number;
 	pageKey?: number;
+	sibling: 'source' | 'dest' | 'both' | 'none';
```
Comment (ryanatkn) [inline src/lib/app/actionTypes.ts:556]: the singular form of this reads a little off to me

File: src/lib/vocab/entity/entityServices.ts [hunk 7]
```diff
@@ -40,13 +45,21 @@ export const ReadEntitiesPaginatedService: ServiceByName['ReadEntitiesPaginated'
...
 		await checkHubAccess(actor, hub_id, repos);
 
-		const ties = await repos.tie.filterBySourceIdPaginated(source_id, pageSize, pageKey);
+		const pageTies = await repos.tie.filterBySourceIdPaginated(source_id, pageSize, pageKey);
+		let siblingTies: Tie[] = [];
+		if (sibling !== 'none' && pageTies.length > 0) {
+			const pageEntityIds = pageTies.map((t) => t.dest_id);
+			siblingTies = await repos.tie.filterSiblingsByEntityId(pageEntityIds, sibling);
+		}
+		const ties = pageTies.concat(siblingTies);
```
Comment (ryanatkn) [inline src/lib/vocab/entity/entityServices.ts:59]: was confusing to see it instantiate an array just to concat it, here's a possible improvement:

```suggestion
		let siblingTies: Tie[] | null = null;
		if (sibling !== 'none' && pageTies.length > 0) {
			const pageEntityIds = pageTies.map((t) => t.dest_id);
			siblingTies = await repos.tie.filterSiblingsByEntityId(pageEntityIds, sibling);
		}
		const ties = siblingTies ? pageTies.concat(siblingTies) : pageTies;
```

if we wanted to be 100% safe but less efficient we could end that line with `pageTies.slice();`

File: src/lib/vocab/entity/entityServices.ts [hunk 8]
```diff
@@ -136,3 +149,21 @@ export const DeleteEntitiesService: ServiceByName['DeleteEntities'] = {
...
+		//TODO make this batchable; need an answer to potential multiHubAccess & rejection question
+		const {actor, entityIds} = params;
+
+		await checkEntityAccess(actor, entityIds, repos);
+
+		const {entities, missing} = await repos.entity.filterByIds(entityIds);
+		if (!missing) {
+			return {ok: true, status: 200, value: {entities}};
+		} else {
+			return {ok: false, status: 404, message: 'Some entities not found'};
```
Comment (ryanatkn) [inline src/lib/vocab/entity/entityServices.ts:166]:
```suggestion
			return {ok: false, status: 404, message: 'failed to find some entities'};
```

File: src/lib/vocab/policy/policyHelpers.server.ts [hunk 9]
```diff
@@ -77,3 +77,23 @@ export const checkEntityOwnership = async (
...
+	entityIds: number[],
+	repos: Repos,
+): Promise<void> => {
+	if (await isPersonaAdmin(actor_id, repos)) {
+		return;
+	}
+
+	const spaces = await repos.space.filterByEntities(entityIds);
+	const assignments = await repos.assignment.filterByPersona(actor_id);
+	const accessableHubIds = assignments.map((a) => a.hub_id);
```
Comment (ryanatkn) [inline src/lib/vocab/policy/policyHelpers.server.ts:92]: maybe leave a todo here for something like `repos.hub.filterByPersona` to optimize this?

File: src/lib/vocab/tie/TieRepo.ts [hunk 10]
```diff
@@ -74,6 +74,37 @@ export class TieRepo extends PostgresRepo {
...
+	/**
+	 * This query, given an array of entities and a couple flags,
+	 * will find all ties pointing at those enties
+	 * all ties pointing away from those enties
+	 * or both
+	 *
+	 * @param entityIds - The entities whose sibling ties you want to find
+	 * @param siblingDest - If enabled, all ties with provided entities as dest are returned
+	 * @param siblingSource - If enabled, all ties with provided entities as source are returned
+	 * @returns - The collection of sibling ties relating to the entityIds (or an empty array)
```
Comment (ryanatkn) [inline src/lib/vocab/tie/TieRepo.ts:86]:
```suggestion
	 * @returns The collection of sibling ties relating to the entityIds (or an empty array)
```

**Review by ryanatkn — APPROVED** [review]
awesome stuff, really fills out querying, I had a couple notes about the API that I think could be improved

File: src/lib/util/query.ts [hunk 11]
```diff
@@ -20,6 +20,7 @@ export interface Query {
 export interface QueryParams {
 	actor: number;
 	source_id: number;
+	sibling: 'dest' | 'source' | 'both' | 'none';
```
Comment (ryanatkn) [inline src/lib/util/query.ts:23]: related: https://github.com/feltjs/felt-server/pull/757/files#r1154923729

File: src/lib/plugins/feltcoop/chat/Chat.svelte (hunk 3 above)
Comment (greatbacon) [reply src/lib/plugins/feltcoop/chat/Chat.svelte:25]: That makes a lot more sense lol

**Review by ryanatkn — APPROVED** [review]

**Comment by ryanatkn** [comment]
CI is failing because that ESLint issue is a false-positive in VSCode, there's some dependency that's not updated, I have the same issue often. Running `gro lint` locally should show there's no issue.

File: src/lib/vocab/actor/actorHelpers.ts [hunk 12]
```diff
@@ -49,8 +49,10 @@ export const checkPersonaName = (name: string): string | null => {
 
 export const lookupPersona = (
 	personaById: Map<ActorId, Readable<ClientActor>>,
-	persona_id: ActorId,
-): Readable<ClientActor> => personaById.get(persona_id) || personaById.get(GHOST_ACTOR_ID)!;
+	// eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
```
Comment (ryanatkn) [inline src/lib/vocab/actor/actorHelpers.ts:52]:
```suggestion
```

see my other comment

**Pull request merged**